dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/validator/relationship/RelationshipValidatorTest.java

```diff
@@ -0,0 +1,191 @@
+
+        validator.validate( reporter, bundle, bundle );
+
+        assertEquals( 1, reporter.getErrors().size(),
+            "Expected just one error when uid not correct, but " + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E1048, "uid: `wrong uid`, has an invalid uid format." );
+    }
+
+    @Test
+    void shouldFailWhenRelationshipFromNotSet()
+    {
+        Relationship relationship = new Relationship();
+        relationship.setRelationship( CodeGenerator.generateUid() );
+        relationship.setTo( new RelationshipItem( "entity", "", "" ) );
+        relationship.setRelationshipType( MetadataIdentifier.ofUid( "uid" ) );
+        when( bundle.getRelationships() ).thenReturn( List.of( relationship ) );
+        when( bundle.getPreheat() ).thenReturn( preheat );
+
+        validator.validate( reporter, bundle, bundle );
+
+        assertEquals( 1, reporter.getErrors().size(),
+            "Expected just one error when from constraint not set, but " + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E1124, "Missing required relationship property: `from`." );
+    }
+
+    @Test
+    void shouldFailWhenRelationshipFromItemTypeNotSet()
+    {
+        String uid = CodeGenerator.generateUid();
+        Relationship relationship = createProgramRelationship( "", uid );
+        when( bundle.getRelationships() ).thenReturn( List.of( relationship ) );
+        when( preheat.getRelationshipType( relationship.getRelationshipType() ) )
+            .thenReturn( createTrackedEntityRelationship() );
+        when( bundle.getPreheat() ).thenReturn( preheat );
+
+        validator.validate( reporter, bundle, bundle );
+
+        assertEquals( 1, reporter.getErrors().size(),
+            "Expected just one error when from item type not set, but " + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E4001, "Relationship Item `from` for Relationship `" + uid
+            + "` is invalid: an Item can link only one Tracker entity." );
+    }
+
+    @Test
+    void shouldFailWhenRelationshipLinksToItself()
+    {
+        String uid = CodeGenerator.generateUid();
+        Relationship relationship = createProgramRelationship( "enrollment", uid );
+        when( bundle.getRelationships() ).thenReturn( List.of( relationship ) );
+        when( preheat.getRelationshipType( relationship.getRelationshipType() ) )
+            .thenReturn( createTrackedEntityRelationship() );
+        when( bundle.getPreheat() ).thenReturn( preheat );
+
+        validator.validate( reporter, bundle, bundle );
+
+        assertEquals( 1, reporter.getErrors().size(), "Expected just one error when relationship links to itself, but "
+            + reporter.getErrors().size() + " found" );
+        assertHasError( reporter, relationship, E4000, "Relationship: `" + uid + "` cannot link to itself" );
```

Comment:
Are we testing the same thing we already tested in the `LinkValidatorTest`?

Is there any interesting "integration" scenario we should test? 🤔 The `RelationshipValidator` puts the individual validators together. It currently runs all of them in a seq so we cannot test that multiple errors are returned to the user within one relationship like when grouping Validators in `all`. I am not sure we need this test class if the individual Validators are well covered.

Reply:
My idea was to test it because of validating the order in which they are run. While it's true the individual test is the same, I wanted to see that since they are all run in seq, there's only one error at a time.

Reply:
I don't think we should duplicate all of these tests in the Validator's own unit test and here. If I break a Validator I will then have 2 failing tests that test exactly the same behavior. Is there one specific scenario that can only be tested in integration like this? If not I don't think we need these tests here which is ok 😄

Reply:
Other than to validate that they are run in sequence, I don't see any other reason to test it. But I don't think we should integration test it when it's something we can unit test.

Reply:
I did not mean with a DB by "integration". To me this test is about testing the integration of all relationship Validators. If there is no special scenario we can only test this way I would delete the tests as they are duplicates. I'll let @enricocolasante way in. If you both feel like they add more value than cost then I am ok 😄

Reply:
I agree that the order is the only thing we are not testing in the validator-specific tests.

So maybe one case that should be tested is to make sure that if we have the same relationship for `from` and `to` but the relationship is invalid, we want the error to be returned to be why the relationship is invalid and not the fact that we don't allow auto relationships.

Or thinking the other way around, what would it happen if we switch the order of LinkValidator and ConstraintEntityValidator?

In any case I would review the tests that we have there, like `shouldFailWhenRelationshipUidDoesNotHaveCorrectFormat` does not add any value to me and I agree with Ivo will just result in a duplicate failure if something goes wrong.

But I wouldn't spend a lot of time on that as it can get really complicated soon, so if you can spot something meaningful that will fail if some order is not respected, go on on that, otherwise, we can remove it.